src/priceClient/priceClient.ts

```diff
@@ -0,0 +1,177 @@
+// of TokenPrice objects, so the address must also be embedded within the TokenPrice object. To
+// avoid storing multiple copies of the same TokenPrice, always use the lower-case variant of the
+// address when performing a lookup into the PriceCache, and substitute the provided address into the
+// TokenPrice object when returning to the caller.
+export type PriceCache = {
+  [address: string]: TokenPrice;
+};
+
+export function msToS(ms: number): number {
+  return Math.floor(ms / 1000);
+}
+
+export class PriceClient implements PriceFeedAdapter {
+  public readonly name: string = "PriceClient";
+  private _maxPriceAge = 300; // seconds
+  protected prices: {
+    [currency: string]: PriceCache;
+  } = {};
+
+  constructor(protected logger: Logger, readonly priceFeeds: PriceFeedAdapter[]) {
+    assert(logger, "No logging instance supplied.");
+    assert(priceFeeds.length > 0, "No price feeds supplied.");
+  }
+
+  get maxPriceAge(): number {
+    return this._maxPriceAge;
+  }
+
+  set maxPriceAge(age: number) {
+    assert(age >= 0);
+    this.logger.debug({
+      at: "PriceClient#maxPriceAge",
+      message: `Setting maxPriceAge (S) ${this._maxPriceAge} => ${age}.`,
+    });
+    this._maxPriceAge = age;
+  }
+
+  listPriceFeeds(): string[] {
+    return this.priceFeeds.map((priceFeed: PriceFeedAdapter) => priceFeed.name);
+  }
+
+  async getPriceByAddress(address: string, currency = "usd"): Promise<TokenPrice> {
+    const tokenPrices = await this.getPricesByAddress([address], currency);
+    return tokenPrices[0];
+  }
+
+  async getPricesByAddress(addresses: string[], currency = "usd"): Promise<TokenPrice[]> {
+    assert(this.priceFeeds.length > 0, "No price feeds are registered.");
+    const priceCache = this.getPriceCache(currency);
+
+    // Determine whether *all* prices are current.
+    const now = msToS(Date.now());
+    const missed: { [address: string]: number } = {};
+    addresses.forEach((address: string) => {
+      const addr = address.toLowerCase();
+      // note: If priceCache[addr] is undefined, it is assigned here.
+      const tokenPrice = (priceCache[addr] = priceCache[addr] ?? ({ price: 0, timestamp: 0 } as TokenPrice));
+      const age = tokenPrice ? now - tokenPrice.timestamp : Number.MAX_SAFE_INTEGER;
+
+      if (age > this.maxPriceAge) {
+        missed[address] = age;
+      }
+    });
+
+    if (Object.keys(missed).length > 0) {
+      const requestAddresses = Object.keys(priceCache);
+      this.logger.debug({
+        at: "PriceClient#getPricesByAddress",
+        message: `${currency.toUpperCase()} cache miss (age > ${this.maxPriceAge} S).`,
+        tokens: missed,
+      });
+      const prices = await this.requestPrices(requestAddresses, currency);
+      if (Object.keys(prices).length === 0) {
+        this.logger.warn({ at: "PriceClient#getPricesByAddress", message: "Failed to update token prices." });
+        // @todo throw ?
+        return [];
+      }
```

Comment:
How could this ever happen? It looks like the requestPrices function always throws if the lookup fails (or returns 0 results), right?

Also, do we prefer to warn or throw an uncaught error? requestPrices throws, but this function just seems to want to warn.

Reply:
Have flip-flopped on this a bit...In conclusion I think it probably makes sense to throw if all lookups failed. That is actually what has been implemented, because there's a test that validates that. So the log message and comment were both redundant here.